Update to latest spec

danielgtaylor opened the pull request:
This PR updates to the latest refract specification, as well as:
1. Convert to ES6 and remove Coffeescript dependencies
2. Use plain old npm as the build system
3. Adds and exposes a preconfigured type registry
4. Adds a `fromCompactRefract` method to read in compact serialized refract elements
5. Renames a few things, adds a few minor things (e.g. StringType#length)

It **does** handle plain js types for the various `push`, `add`, `set`, etc methods as well as already refracted types so long as they inherit from `ElementType`. It **does not** do any more work for error handling.

This is merging the work from apiaryio/fury.js#16 back into the minim project. The new SDK API in Fury.js is depending on these changes.

Sorry it's so large :weary:

danielgtaylor committed:
Update to latest spec

danielgtaylor commented on README.md:
````diff
@@ -142,7 +148,7 @@ This is a type for representing string values.
 
 ```javascript
-var stringType = new minim.StringType('foobar');
+var stringType = new minim.StringType({}, {}, 'foobar');
````
The need to include meta/attributes in the constructor was mainly for consistency sake, but it makes for a worse experience in my opinion. Ideas?

smizell replied:
One thing we could do is do this backwards:

``` javascript
var stringType = new minim.StringType('foobar', {}, {});
```

Where the first argument was the content of the element.

danielgtaylor replied:
I've updated this in the latest commit. The order is now `ElementType(element, content, meta, attributes)` and the other types don't require an element, e.g. `StringType('some string');` just like before. The important note is that even the base element type has `content` as the second argument for consistency sake.

danielgtaylor committed:
Update tested node versions

danielgtaylor commented:
Here's a case that this PR does not provide any conveniences for yet:

https://github.com/refractproject/refract-spec/blob/master/namespaces/mson-namespace.md#referencing--expansion

Edit: After spending some time with this today I think we may need to change how the type registry works. When loading an element type which hasn't been encountered before and is not present in the registry we should probably add it to the registry so that inheritance works. If `A` is an object and `B` is an `A`, then `B` should be loaded as an object and its content treated as a collection. With the code as-is in the PR it will ignore the content, which is a problem.

danielgtaylor committed:
Change constructor argument order; default to const

danielgtaylor committed:
Make collections iterable

danielgtaylor committed:
Use babel-runtime instead of core-js

smizell commented on README.md:
````diff
@@ -76,18 +82,18 @@ If the JSON above is used, it can be converted back to Minim types to make a rou
 ```
 
-### Element Methods
+### Element attributes
````
Capitalize "Attributes"

smizell commented on README.md:
```diff
@@ -76,18 +82,18 @@ If the JSON above is used, it can be converted back to Minim types to make a rou
+Each Minim type provides the following attributes:
 
-#### elementType
+- element: `array` (string) - The name of the element type.
```
Should `array` be here?

smizell commented on README.md:
```diff
@@ -206,6 +212,18 @@ var value = boolType.get() // get() returns 4
 
 This is a type for representing arrays.
 
+##### iteration
```
I'd capitalize this.

smizell commented on README.md:
````diff
@@ -317,7 +335,37 @@ The `values` method returns an array of keys.
+```js
+const objectType = new minim.ObjectType({ foo: 'bar' });
+
+for (let [key, value] of objectType.items()) {
````
:+1:

smizell commented on README.md:
```diff
@@ -317,7 +335,37 @@ The `values` method returns an array of keys.
+var minim = require('minim');
+
+// Register your custom type
+minim.TypeRegistry.elementMap.category = minim.ArrayType;
```
Regarding this, it almost feels like there needs to be something for registering.

``` javascript
minim.TypeRegistry.register('category', minim.ArrayType);
```

smizell commented on index.js:
```diff
@@ -1 +1,6 @@
+module.exports = _.extend(
+  require('./lib/primitives'),
+  require('./lib/convert')
+);
```
Does this mutate what is exprted from `./lib/primitives`? Is that desired? It almost seems like you need:

``` javascript
module.exports = _.extend(
  {},
  require('./lib/primitives'),
  require('./lib/convert')
);
```

smizell commented on src/convert.es6:
```diff
@@ -0,0 +1,53 @@
+    // Fall back to the base element. We may not know what
+    // to do with the `content`, but downstream software
+    // may know.
+    ElementClass = ElementType;
```
For the sake of trying not to mutate values, you could instead:

``` javascript
return ElementType
```

smizell commented on src/primitives.es6:
```diff
@@ -0,0 +1,339 @@
+    return this.content;
+  }
+
+  toRefract(options={}) {
```
I think it would be good to start following this:
https://github.com/airbnb/javascript

If you want to try to do it in this PR, feel free, but we can also merge this and try to do it in another PR.

For example, this line should be:

``` javascript
toRefract(options = {}) {
```

smizell replied:
Just to point it out, note the spacing around the equals sign.

danielgtaylor replied:
Ah, sorry, I'm coming from Python and PEP8 land :-)

smizell replied:
I thought I saw some Python with the `for [key, value] in something.items()` :)

smizell commented on src/primitives.es6:
```diff
@@ -0,0 +1,339 @@
+    return this.content.length;
+  }
+
+  get [Symbol.iterator]() {
```
Being relatively new to ES6, a comment here would help me know what's going on here.

smizell commented on src/primitives.es6:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  set(name, value) {
+    const location = this.content.map(i => i.meta.name).indexOf(name);
```
Would something other than `i` be better here?

smizell commented on src/convert.es6:
```diff
@@ -0,0 +1,53 @@
+export function convertFromRefract(dom) {
+  let ElementClass = getElementClass(dom.element);
+  return new ElementClass().fromRefract(dom);
+}
```
Just getting this idea out, but it may be good for us to come up with better names than "long-form" and "DOM" here. Something we stay consistent on throughout the spec and libraries.

smizell commented:
@danielgtaylor I have added a few comments throughout this PR, but I like what you'd done. This is all good stuff! We should try to get this merged ASAP once you have a look at the comments.

smizell commented on src/primitives.es6:
```diff
@@ -0,0 +1,339 @@
+    const newArray = new Collection();
+    newArray.content = this.findElements(condition, options);
+    return newArray;
+  }
```
I noticed in the reveal today, you used `findElements` instead of `find`. Are you wanting to move away from `find` or did `findElements` work better?

Just as a note, if you're wanting to move away from it, the README still shows `find` examples.

smizell replied:
I'm wondering if `find` should be recursive by default, and then have something like `children` which is not recursive. What do you think?

smizell commented on src/primitives.es6:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  findElements(condition, options={}) {
+    const recursive = options.recursive === undefined ? false : options.recursive;
```
Do we trust the user, or do you want to convert the recursive value given to boolean?

``` javascript
!!options.recursive
```

Just a question that popped up.

smizell commented on src/primitives.es6:
```diff
@@ -0,0 +1,339 @@
+    //       for non-refracted elements as they get copied anyway, but
+    //       if the input is already refracted and we add it to multiple
+    //       objects with a different name suddenly we have a problem.
+    //       We have the same problem in the constructor above.
```
Good point. I'm wondering if we can handle this someway internally with a special kind of element wrapper. It would simply wrap an element in a property name.

Just throwing out ideas. Would be nice to have our cake and eat it, too.

danielgtaylor committed:
Switch to Airbnb style linter config

danielgtaylor committed:
Code review feedback.

- Adhere to Airbnb ES6 style guide (in `src` folder)
- Default `find` to be recursive.
- Add `children` which is non-recursive.
- Make type registry a proper class, add tests.
- Minor capitalization fixes & other minor fixes.

danielgtaylor commented:
@smizell I've addressed most of the feedback in the latest push. Please have a look and let me know if you have any other comments.

danielgtaylor committed:
Temporarily make no-shadow eslint rule a warning

danielgtaylor commented:
Had to disable the `no-shadow` rule temporarily until eslint/eslint#2547

danielgtaylor committed:
Fix attribute de/serialization

smizell commented:
:+1: Looks great!
